mantidimaging/gui/windows/spectrum_viewer/model.py

```diff
@@ -28,50 +34,69 @@ def set_stack(self, stack: ImageStack) -> None:
         self._stack = stack
         self.tof_range = (0, stack.data.shape[0] - 1)
         height, width = self.get_image_shape()
-        self.roi_range = SensibleROI.from_list([0, 0, width, height])
+        self.set_roi("all", SensibleROI.from_list([0, 0, width, height]))
+        self.set_roi("roi", SensibleROI.from_list([0, 0, width, height]))
 
     def set_normalise_stack(self, normalise_stack: ImageStack) -> None:
         self._normalise_stack = normalise_stack
 
+    def set_roi(self, roi_name: str, roi: SensibleROI):
+        self._roi_ranges[roi_name] = roi
+
+    def get_roi(self, roi_name: str):
+        return self._roi_ranges[roi_name]
+
     def get_averaged_image(self) -> Optional['np.ndarray']:
         if self._stack is not None:
             tof_slice = slice(self.tof_range[0], self.tof_range[1] + 1)
             return self._stack.data[tof_slice].mean(axis=0)
         else:
             return None
 
-    def get_spectrum(self) -> Optional['np.ndarray']:
+    @staticmethod
+    def get_stack_spectrum(stack: ImageStack, roi: SensibleROI):
+        left, top, right, bottom = roi
+        roi_data = stack.data[:, top:bottom, left:right]
+        return roi_data.mean(axis=(1, 2))
+
+    def get_spectrum(self, roi_name: str, mode: SpecType) -> Optional['np.ndarray']:
```

Comment:
A thought on the issue of mapping back and forth between a spectrum type, I wonder if we could resolve that by splitting up this method? So you would have `get_sample_spectrum`, `get_open_spectrum` and `get_sample_norm_spectrum`, keeping the `get_stack_spectrum` method to use within each of those. I think this would have the knock on effect that the presenter could just call the relevant method instead of having to pass in a type. In the places where the presenter needs to establish if we're normalising or not, could the view provide a method to get the status of the normalise checkbox and then the presenter could call that when it needs to find out if we're normalising instead of holding the state itself?

Reply:
Or the presenter could hold the boolean state that gets passed to `handle_enable_normalised` if we prefer not to keep looking it up.

Reply:
Maybe, though it might move more complexity into the presenter. I think I'll come back to this later.